### PR TITLE
Use a contextual UnderscoreToken for the discard pattern.

### LINKED ISSUE #27750: Should treat `_` as a contextual keyword
In the `features/recursive_patterns` branch, we have started to treat `_` as a contextual keyword. However, there are a few places that we don't, such as this:

``` c#
            if (tk == SyntaxKind.IdentifierToken && this.CurrentToken.Text == "_")
            {
                // In a pattern, we reserve `_` as a discard. It cannot be used (with that spelling) as the
                // type of a declaration or recursive pattern, nor as a type in an is-type expression starting
                // in C# 8. The binder will give a diagnostic if
                // there is a usable symbol in scope by that name. You can always escape it, using `@_`.
                return _syntaxFactory.DiscardPattern(this.EatToken(SyntaxKind.IdentifierToken));
            }
```

We should be consistent, probably by treating it as a contextual keyword here too.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #27750

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 1]
```diff
@@ -1758,7 +1758,7 @@
   <Node Name="DiscardPatternSyntax" Base="PatternSyntax">
     <Kind Name="DiscardPattern" />
     <Field Name="UnderscoreToken" Type="SyntaxToken">
-      <Kind Name="IdentifierToken"/>
+      <Kind Name="UnderscoreToken"/>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: ><Kind Name="UnderscoreToken"/> [](start = 6, length = 30)

Is this a breaking change?

**Comment by AlekseyTs** [comment]
It looks like there are some test failures

**Comment by AlekseyTs** [comment]
It looks like this change is not covered by targeted tests

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: No. The discard pattern is new.

**Comment by gafter** [comment]
> It looks like there are some test failures

Yes, some intermittent tests unrelated to these changes failed. Retesting passed.

**Comment by gafter** [comment]
@cston @agocke Could you please review this pattern-matching change?

**Comment by gafter** [comment]
@AlekseyTs This PR is now ready for review if you'd like to look at it.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 2]
```diff
@@ -6340,6 +6335,15 @@ static void M(out int x)
         M(out _);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Should these be calling `M1`?

Comment (agocke) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Yeah, this looks wrong.

Comment (gafter) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Agreed; fixed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 3]
```diff
@@ -165,7 +165,8 @@ private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol in
             out bool wasExpression)
         {
             if (innerExpression.Kind() == SyntaxKind.IdentifierName &&
-                ((IdentifierNameSyntax)innerExpression).Identifier.Text == "_")
+                ((IdentifierNameSyntax)innerExpression).Identifier.Text == "_" &&
+                Compilation.LanguageVersion >= MessageID.IDS_FeatureRecursivePatterns.RequiredVersion())
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: In particular, it doesn't look like existing tests were affected by this additional condition. 

---
In reply to: [211746992](https://github.com/dotnet/roslyn/pull/29290#discussion_r211746992) [](ancestors = 211746992)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs]: I've updated `UnderscoreDeclaredAndDiscardPattern_01` to show the effect of this change.

**Comment by gafter** [comment]
@agocke @AlekseyTs Do you have any further comments?

**Pull request merged**